pennylane/beta/tapes/tape.py

```diff
@@ -936,3 +936,387 @@ def execute_device(self, params, device):
+        """
+        if params is None:
+            params = np.array(self.get_parameters())
+
+        order = options.get("order", 1)
+        h = options.get("h", 1e-7)
+
+        shift = np.zeros_like(params)
+        shift[idx] = h
+
+        if order == 1:
+            # Forward finite-difference.
+            # Check if the device has already be pre-computed with
+            # unshifted parameter values, to avoid redundant evaluations.
+            y0 = options.get("y0", None)
+
+            if y0 is None:
+                y0 = np.asarray(self.execute_device(params, device))
+
+            y = np.array(self.execute_device(params + shift, device))
+            return (y - y0) / h
+
+        if order == 2:
+            # central finite difference
+            shift_forward = np.array(self.execute_device(params + shift / 2, device))
+            shift_backward = np.array(self.execute_device(params - shift / 2, device))
+            return (shift_forward - shift_backward) / h
+
+        raise ValueError("Order must be 1 or 2.")
+
+    def device_pd(self, device, params=None, **options):
+        """Evaluate the gradient of the tape with respect to
+        a single trainable tape parameter by querying the provided device.
+
+        Args:
+            device (~.Device, ~.QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+        """
+        # pylint:disable=unused-argument
+        if params is None:
+            params = np.array(self.get_parameters())
+
+        saved_parameters = self.get_parameters()
+
+        # temporarily mutate the in-place parameters
+        self.set_parameters(params)
+
+        # TODO: modify devices that have device Jacobian methods to
+        # accept the quantum tape as an argument
+        jac = device.jacobian(self)
+
+        # restore original parameters
+        self.set_parameters(saved_parameters)
+        return jac
+
+    def analytic_pd(self, idx, device, params=None, **options):
+        """Evaluate the gradient of the tape with respect to
+        a single trainable tape parameter using an analytic method.
+
+        Args:
+            idx (int): trainable parameter index to differentiate with respect to
+            device (~.Device, ~.QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+
+        Returns:
+            array[float]: 1-dimensional array of length determined by the tape output
+                measurement statistics
+        """
+        raise NotImplementedError
```

Comment:
Note: we explicitly _don't_ test `NotImplementedError` in our `.coveragerc` file, as the tests for this are trivial. This is codecov playing it (it seems to have stopped reading the `.coveragerc` info)